substrate/client/rpc-spec-v2/src/common/connections.rs

```diff
@@ -0,0 +1,228 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use jsonrpsee::ConnectionId;
+use parking_lot::Mutex;
+use std::{
+	collections::{HashMap, HashSet},
+	sync::Arc,
+};
+
+/// Limit the RPC functionality to a single connection.
+#[derive(Default, Clone)]
+pub struct RpcConnections {
+	/// The number of tokens that can be registered for each connection.
+	capacity: usize,
+	/// Map the connecton ID to a set of tokens.
+	data: Arc<Mutex<HashMap<ConnectionId, ConnectionData>>>,
+}
+
+#[derive(Default)]
+struct ConnectionData {
+	/// The total number of tokens.
+	///
+	/// # Note
+	///
+	/// Because a pending subscription sink does not expose the future subscription ID,
+	/// we cannot register a token before the pending subscription is accepted.
+	/// This variable ensures that we have enough capacity to register a token, after
+	/// the subscription is accepted. Otherwise, a jsonrpc error object should be returned.
+	num_tokens: usize,
+	/// Active registered tokens.
+	tokens: HashSet<String>,
+}
+
+impl RpcConnections {
+	/// Constructs a new instance of [`RpcConnections`].
+	pub fn new(capacity: usize) -> Self {
+		RpcConnections { capacity, data: Default::default() }
+	}
+
+	/// Reserve space for a token.
+	pub fn reserve_token(&self, connection_id: ConnectionId) -> Option<ReservedConnectionToken> {
+		let mut data = self.data.lock();
+
+		let entry = data.entry(connection_id).or_insert_with(ConnectionData::default);
+		if entry.num_tokens >= self.capacity {
+			return None;
+		}
+		entry.num_tokens = entry.num_tokens.saturating_add(1);
+
+		Some(ReservedConnectionToken { connection_id, rpc_connections: Some(self.clone()) })
+	}
+
+	/// Gives back the reserved space before the token is registered.
+	///
+	/// # Note
+	///
+	/// This may happen if the pending subscription cannot be accepted (unlikely).
+	fn unreserve_token(&self, connection_id: ConnectionId) {
+		let mut data = self.data.lock();
+
+		let entry = data.entry(connection_id).or_insert_with(ConnectionData::default);
+		entry.num_tokens = entry.num_tokens.saturating_sub(1);
+	}
+
+	/// Register a token for the given connection.
+	///
+	/// Users should call [`Self::reserve_token`] before calling this method.
+	fn register_token(&self, connection_id: ConnectionId, token: String) {
+		let mut data = self.data.lock();
+
+		let entry = data.entry(connection_id).or_insert_with(ConnectionData::default);
+		// Should be already checked `Self::reserve_token`.
+		if entry.tokens.len() >= self.capacity {
+			return;
+		}
+
+		entry.tokens.insert(token);
```

Comment:
we assume that subscription ID is unique so should be unlikely for duplicate subscription IDs here